bytecomp/translcore.ml

```diff
@@ -848,20 +850,35 @@ and transl_exp0 e =
       let kind = array_kind e in
       let ll = transl_list expr_list in
       begin try
+        (* For native code the decision as to which compilation strategy to
+           use is made later.  This enables the Flambda passes to lift certain
+           kinds of array definitions to symbols. *)
         (* Deactivate constant optimization if array is small enough *)
-        if List.length ll <= 4 then raise Not_constant;
-        let cl = List.map extract_constant ll in
-        let master =
-          match kind with
-          | Paddrarray | Pintarray ->
-              Lconst(Const_block(0, cl))
-          | Pfloatarray ->
-              Lconst(Const_float_array(List.map extract_float cl))
-          | Pgenarray ->
-              raise Not_constant in             (* can this really happen? *)
-        Lprim(Pccall prim_obj_dup, [master])
+        if List.length ll <= use_dup_for_constant_arrays_bigger_than
+        then begin
+          raise Not_constant
+        end;
+        (* We cannot currently lift [Pintarray] arrays safely in Flambda
+           because [caml_modify] might be called upon them (e.g. from
+           code operating on polymorphic arrays, or functions such as
+           [caml_array_blit]. *)
+        if !Clflags.native_code && kind = Pfloatarray then
```

Comment:
I will improve this comment.  The fundamental point is that if the array (defined by making a mutable copy of an immutable array) is lifted to toplevel, Flambda can do the caml_obj_dup at compile time.  As such, the mutable array ends up outside the heap, and might (unless it is a float array) have caml_modify called upon it, which would fail in such circumstance.